### DISCUSSION TITLE
H4 - custom keyboard - bug in TCA8418 library and solution

### DISCUSSION THREAD

**Discussion by ilmarietto123** [kachurovskiy/nanoels#234] [General]
If someone like me want to make a custom keyboard with different layout for nanoELS H4, I hope this can be helpfull.
I found out there is a known bug in the Adafruit_TCA8418 library version 1.0.2.
If you try to use the 9th and 10th columns of the chip (named in datasheet "C8" and "C9"), you get a double/wrong replay.
You can find the issue here:

[https://github.com/adafruit/Adafruit_TCA8418/issues/4](url)

Following the Adafruit forum link in the issue, there is a working solution, modifing the library file:
"C:\Users\<your user profile>\Documents\Arduino\Libraries\Adafruit_TCA8418\Adafruit_TCA8418.cpp"
from line 72 to 75.

Change lines from this:
```
  //  add all pins to key events
  writeRegister(TCA8418_REG_GPI_EM_1, 0xFF);
  writeRegister(TCA8418_REG_GPI_EM_2, 0xFF);
  writeRegister(TCA8418_REG_GPI_EM_3, 0xFF);
```
to this:
```
  //  remove all pins to key events
  writeRegister(TCA8418_REG_GPI_EM_1, 0x00);
  writeRegister(TCA8418_REG_GPI_EM_2, 0x00);
  writeRegister(TCA8418_REG_GPI_EM_3, 0x00);
```

Remember to modify also H4.ino accordingly to the size of your matrix keyboard at line [1472](https://github.com/kachurovskiy/nanoels/blob/8cc4b13d93870869a071ace68bb6aeda617e1e79/h4/h4.ino#L1742C1-L1742C25):
```
keypad.matrix(7, 7);
```
Example: my keyboard is 6 rows and 10 columns:
```
keypad.matrix(6, 10);
```

Then you can change the #define value of your buttons in the code [from line 112 to 157](https://github.com/kachurovskiy/nanoels/blob/8cc4b13d93870869a071ace68bb6aeda617e1e79/h4/h4.ino#L119-L157) with the values indicated in [TCA8418 datasheet table 1](https://www.ti.com/document-viewer/TCA8418/datasheet#key-event-table-scps2224386/SCPS2224386).

Thanks @kachurovskiy for your great project!

**Comment by kachurovskiy** [comment]
@ilmarietto123 thanks for sharing this workaround!

**Comment by kotovasia5120** [comment]
a little help on the buttons from the firmware.
about how they are located in the PDF file.

`//code keypad/* C0 C1 C2 C3 C4 C5 C6R0 1 2 3 4 5 6 7R1 11 12 13 14 15 16 17 R2 21 22 23 24 25 26 27R3 31 32 33 34 35 36 37R4 41 42 43 44 45 46 47R5 51 52 53 54 55 56 57R6 61 62 63 64 65 66 67*/

define B_ON 17 //START
#define B_OFF 27 //STOP
#define B_STOPL 7 //LLEFT
#define B_STOPR 15 //LRIGHT
#define B_STOPU 6 //LUP
#define B_STOPD 16 //LDOWN
#define B_DISPL 14 //M5
#define B_STEP 24 //M4
#define B_SETTINGS 34 //M3
#define B_MEASURE 54 //M1
#define B_REVERSE 44 //M2
#define B_BACKSPACE 32 //BS
#define B_MODE_GEARS 42 //F1
#define B_MODE_TURN 52 //F2
#define B_MODE_FACE 62 //F3
#define B_MODE_CONE 3 //F4
#define B_MODE_CUT 13 //F5
#define B_MODE_THREAD 23 //F6
#define B_MODE_OTHER 33 //F7
#define B_X 53 //M8
#define B_Z 43 //M9
#define B_A 4 //M6
#define B_B 63 //M7`
